### PR TITLE
Fix umount code insertion position

### PR DESCRIPTION
**Pull request by Fyg369**
## Type
- [ ] bug fix 

- [ ] feature add 

- [ ] other 

## Checkbox
- [ ] do not have break changes 

- [ ] normal operation will not be affected after modification 

## Linked issues
<!-- if want to fix it, try "fixes: #13">

### CODE REVIEW

File: kernelsu/patch.sh [hunk 1]
```diff
@@ -65,7 +65,7 @@ for i in "${patch_files[@]}"; do
         ;;
 
     fs/namespace.c)
-        sed -i '/int ksys_umount(char __user \*name, int flags)/i \
+        sed -i '/may_mandlock(void)/,/^}/ { /^}/a \
```
Comment (dabao1955) [inline kernelsu/patch.sh:68]: 使用grep检测到没有这个函数然后sed可能会更好一些

**Review by dabao1955 — CHANGES_REQUESTED** [review]

File: kernelsu/patch.sh (hunk 1 above)
Comment (Fyg369) [reply kernelsu/patch.sh:68]: > 使用grep检测到没有这个函数然后sed可能会更好一些

这样修改？
fs/namespace.c)
    can_umount_code='#ifdef CONFIG_KSU
static int can_umount(const struct path *path, int flags)
{
    struct mount *mnt = real_mount(path->mnt);

    if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))
        return -EINVAL;
    if (!may_mount())
        return -EPERM;
    if (path->dentry != path->mnt->mnt_root)
        return -EINVAL;
    if (!check_mnt(mnt))
        return -EINVAL;
    if (mnt->mnt.mnt_flags & MNT_LOCKED) /* Check optimistically */
        return -EINVAL;
    if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))
        return -EPERM;
    return 0;
}

int path_umount(struct path *path, int flags)
{
    struct mount *mnt = real_mount(path->mnt);
    int ret;

    ret = can_umount(path, flags);
    if (!ret)
        ret = do_umount(mnt, flags);

    /* we must not call path_put() as that would clear mnt_expiry_mark */
    dput(path->dentry);
    mntput_no_expire(mnt);
    return ret;
}
#endif'

    if grep -q "may_mandlock(void)" "$i"; then
        sed -i "/may_mandlock(void)/,/^}/ { /^}/a\\$can_umount_code" "$i"
    else
        sed -i "/int ksys_umount(char __user \*name, int flags)/i\\$can_umount_code" "$i"
    fi
    ;;

Comment (Fyg369) [reply kernelsu/patch.sh:68]: > 测试过可以就行

经过测试sed命令无法处理这种插入多行代码的变量，大佬有什么好的建议吗？

**Comment by Fyg369** [comment]
试了很多方法都不是很理想，但是用sed命令插入显的代码太冗长了

    fs/namespace.c)
        if grep -q "may_mandlock(void)" "$i"; then
            sed -i '/may_mandlock(void)/,/^}/ { /^}/a \
#ifdef CONFIG_KSU\
static int can_umount(const struct path *path, int flags)\
{\
    struct mount *mnt = real_mount(path->mnt);\
\
    if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))\
        return -EINVAL;\
    if (!may_mount())\
        return -EPERM;\
    if (path->dentry != path->mnt->mnt_root)\
        return -EINVAL;\
    if (!check_mnt(mnt))\
        return -EINVAL;\
    if (mnt->mnt.mnt_flags & MNT_LOCKED) /* Check optimistically */\
        return -EINVAL;\
    if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))\
        return -EPERM;\
    return 0;\
}\
\
int path_umount(struct path *path, int flags)\
{\
    struct mount *mnt = real_mount(path->mnt);\
    int ret;\
\
    ret = can_umount(path, flags);\
    if (!ret)\
        ret = do_umount(mnt, flags);\
\
    /* we must not call path_put() as that would clear mnt_expiry_mark */\
    dput(path->dentry);\
    mntput_no_expire(mnt);\
    return ret;\
}\
#endif
}' fs/namespace.c
        else
            sed -i '/int ksys_umount(char __user \*name, int flags)/i \
#ifdef CONFIG_KSU\
static int can_umount(const struct path *path, int flags)\
{\
    struct mount *mnt = real_mount(path->mnt);\
\
    if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))\
        return -EINVAL;\
    if (!may_mount())\
        return -EPERM;\
    if (path->dentry != path->mnt->mnt_root)\
        return -EINVAL;\
    if (!check_mnt(mnt))\
        return -EINVAL;\
    if (mnt->mnt.mnt_flags & MNT_LOCKED) /* Check optimistically */\
        return -EINVAL;\
    if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))\
        return -EPERM;\
    return 0;\
}\
\
int path_umount(struct path *path, int flags)\
{\
    struct mount *mnt = real_mount(path->mnt);\
    int ret;\
\
    ret = can_umount(path, flags);\
    if (!ret)\
        ret = do_umount(mnt, flags);\
\
    /* we must not call path_put() as that would clear mnt_expiry_mark */\
    dput(path->dentry);\
    mntput_no_expire(mnt);\
    return ret;\
}\
#endif\
' fs/namespace.c
        fi
        ;;

**Comment by dabao1955** [comment]
> 试了很多方法都不是很理想，但是用sed命令插入显的代码太冗长了
> 
>     fs/namespace.c)
>         if grep -q "may_mandlock(void)" "$i"; then
>             sed -i '/may_mandlock(void)/,/^}/ { /^}/a \
> #ifdef CONFIG_KSU\
> static int can_umount(const struct path *path, int flags)\
> {\
>     struct mount *mnt = real_mount(path->mnt);\
> \
>     if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))\
>         return -EINVAL;\
>     if (!may_mount())\
>         return -EPERM;\
>     if (path->dentry != path->mnt->mnt_root)\
>         return -EINVAL;\
>     if (!check_mnt(mnt))\
>         return -EINVAL;\
>     if (mnt->mnt.mnt_flags & MNT_LOCKED) /* Check optimistically */\
>         return -EINVAL;\
>     if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))\
>         return -EPERM;\
>     return 0;\
> }\
> \
> int path_umount(struct path *path, int flags)\
> {\
>     struct mount *mnt = real_mount(path->mnt);\
>     int ret;\
> \
>     ret = can_umount(path, flags);\
>     if (!ret)\
>         ret = do_umount(mnt, flags);\
> \
>     /* we must not call path_put() as that would clear mnt_expiry_mark */\
>     dput(path->dentry);\
>     mntput_no_expire(mnt);\
>     return ret;\
> }\
> #endif
> }' fs/namespace.c
>         else
>             sed -i '/int ksys_umount(char __user \*name, int flags)/i \
> #ifdef CONFIG_KSU\
> static int can_umount(const struct path *path, int flags)\
> {\
>     struct mount *mnt = real_mount(path->mnt);\
> \
>     if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))\
>         return -EINVAL;\
>     if (!may_mount())\
>         return -EPERM;\
>     if (path->dentry != path->mnt->mnt_root)\
>         return -EINVAL;\
>     if (!check_mnt(mnt))\
>         return -EINVAL;\
>     if (mnt->mnt.mnt_flags & MNT_LOCKED) /* Check optimistically */\
>         return -EINVAL;\
>     if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))\
>         return -EPERM;\
>     return 0;\
> }\
> \
> int path_umount(struct path *path, int flags)\
> {\
>     struct mount *mnt = real_mount(path->mnt);\
>     int ret;\
> \
>     ret = can_umount(path, flags);\
>     if (!ret)\
>         ret = do_umount(mnt, flags);\
> \
>     /* we must not call path_put() as that would clear mnt_expiry_mark */\
>     dput(path->dentry);\
>     mntput_no_expire(mnt);\
>     return ret;\
> }\
> #endif\
> ' fs/namespace.c
>         fi
>         ;;

那么patch呢

**Comment by Fyg369** [comment]
重修修改了一下，在乌班图本地测试正常，已提交修改，看下是否还有其它问题需要修改

**Comment by Fyg369** [comment]
已在乌班图本地测试没问题，大佬看下是否还有其它要修改的

**Comment by dabao1955** [comment]
> 已在乌班图本地测试没问题，大佬看下是否还有其它要修改的

如果4.4,4.9,4.14,4.19,5.4测试过都没问题就行

**Comment by dabao1955** [comment]
> > > 已在乌班图本地测试没问题，大佬看下是否还有其它要修改的
> > 
> > 如果4.4,4.9,4.14,4.19,5.4测试过都没问题就行
> 
> 插入的代码其实没做太大改变，只是添加了判断，根据判断插入相应位置的代码，至于下图的代码是为了适应两个不同位置，执行修改删除补丁脚本花括号而已，因为这两个插入位置有一个花括号不兼容。
> 
> 
> 
> 
> 

ok

**Review by dabao1955 — APPROVED** [review]

**Comment by Fyg369** [comment]
分支合并又出现Bug了

**Pull request merged**